### PR TITLE
DSP clock switching refining

### PR DESCRIPTION
**Pull request by keyonjie**
1. add notification event for the clock switching.
2. use the event to notify KPB to start draining.

### CODE REVIEW

File: src/include/sof/lib/clk.h [hunk 1]
```diff
@@ -21,6 +21,7 @@ struct timer;
 #define CLOCK_NOTIFY_POST	1
 
 struct clock_notify_data {
+	uint32_t clk_id;
```
Comment (lgirdwood) [inline src/include/sof/lib/clk.h:24]: This is getting more complex here, could you add some inline comments for each field here.

File: src/schedule/ll_schedule.c [hunk 2]
```diff
@@ -536,6 +536,9 @@ static void ll_scheduler_notify(void *arg, enum notify_id type, void *data)
 	struct clock_notify_data *clk_data = data;
 	uint32_t flags;
 
+	if (clk_data->clk_id != sch->domain->clk)
```
Comment (lgirdwood) [inline src/schedule/ll_schedule.c:539]: can you comment what we are doing here.

File: src/audio/kpb.c [hunk 3]
```diff
@@ -944,6 +944,41 @@ static void kpb_event_handler(void *arg, enum notify_id type, void *event_data)
...
 }
 
+/**
+ * \brief cpu clock event dispatcher.
+ * \param[in] arg - KPB component internal data.
+ * \param[in] type - notification type
+ * \param[in] event_data - event specific data.
+ * \return none.
+ */
+static void kpb_drain_notify(void *arg, enum notify_id type, void *data)
```
Comment (lgirdwood) [inline src/audio/kpb.c:954]: I think this has to be
```c
pm_runtime_get(clock_id, min_freq);
```
to tell the core this is the min clock we need and then we ```put``` it when we are finished.

**Review by lgirdwood — CHANGES_REQUESTED** [review]
We need to make the clock request a reference count for the "min required frequency", clock can be lowered when there are no users of this frequency to next lowest (and so on).

**Comment by keyonjie** [comment]
@lgirdwood @mwasko since our ll_scheduler is using wall clock which actually never change its frequency so no notification is needed, and the notification for KPB looks not an urgent task now, let me close this PR now, we can reopen it if needed in someday.

**Pull request closed without merging**